src/Data/List/Linear.hs

```diff
@@ -0,0 +1,346 @@
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+
+{-|
+Linear versions of 'Data.List' functions.
+
+This module only contains minimal amount of documentation; consult the
+original "Data.List" module for more detailed information.
+-}
+module Data.List.Linear
+  ( -- * Basic functions
+    (++)
+  , map
+  , filter
+  , NonLinear.head
+  , uncons
+  , NonLinear.tail
+  , NonLinear.last
+  , NonLinear.init
+  , reverse
+  , length
+  , NonLinear.null
+    -- * Extracting sublists
+  , take
+  , drop
+  , splitAt
+  , span
+  , partition
+  , takeWhile
+  , dropWhile
+  , NonLinear.find
+  , intersperse
+  , intercalate
+  , transpose
+  -- * Folds
+  , foldl
+  , foldl'
+  , foldl1
+  , foldl1'
+  , foldr
+  , foldr1
+  , foldMap
+  , foldMap'
+  -- * Special folds
+  , concat
+  , concatMap
+  , and
+  , or
+  , any
+  , all
+  , sum
+  , product
+  -- * Building lists
+  , scanl
+  , scanl1
+  , scanr
+  , scanr1
+  , repeat
+  , replicate
+  , cycle
+  , iterate
+  , unfoldr
+  -- * Zipping lists
+  , zip
+  , zip'
+  , zip3
+  , zipWith
+  , zipWith'
+  , zipWith3
+  , unzip
+  , unzip3
+  ) where
+
+import qualified Unsafe.Linear as Unsafe
```

Comment:
No.......... I had hoped this wasn't going to be here. But, it might all work out :crossed_fingers:

Reply:
I agree that currently this module isn't look very safe since we can not ensure that none of those unsafe casts actually drop/duplicate values.

However, `Data.List` has a lot of tricks to make things fast; and implementing them here would be more work. I think this matches your sentence here:

> We should only use Unsafe when there's a non-linear optimized implementation already in base that we want to copy.

(Also, it is likely that the `unsafeCoerce`'s there probably get in the way of many optimisations so I do not think this will be as fast as the `Data.List` even in this shape).

Reply:
I might have looked at the wrong files but when I looked at what's underneath `Data.List` for some of the coerced functions I didn't find optimized versions. In any case, not something we need to worry about right now. I'm not worried about safety, since most of these look like they could have linear implementations. Still, we should have an issue to double check.

Reply:
Oh yes, I did not make that choice per-function basis. I just looked `Data.List` overall, saw some gnarly bits and decided to implement this module by coerces. You are probably right that some of the functions would be easy to duplicate here. I'll create an issue.